File: java/src/dataStructures/LinkedList.java
package dataStructures;

public class LinkedList<T> {
	public static class Node<T> {
		private T value;
		private Node<T> previous;
		private Node<T> next;

		public Node(T value) {
			this.value = value;
		}
		
		public T getValue(){
			return value;
		}

		public void setValue(T value){
			this.value = value;
		}

		public Node<T> getNext(){
			return next;
		}
		
		public void setNext(Node<T> next){
			this.next = next;
		}
		
		public void setNext(T value){
			Node<T> next = new Node<T>(value);
			setNext(next);
		}
		
		public boolean hasNext(){
			return next != null;
		}

		public Node<T> getPrevious(){
			return previous;
		}
		
		public void setPrevious(Node<T> previous){
			this.previous = previous;
		}
		
		public boolean hasPrevious(){
			return previous != null;
		}
		
		public void setPrevious(T value){
			Node<T> previous = new Node<T>(value);
			setNext(previous);
		}
		
		public String toString(){
			return value.toString();
		}
	}

	protected Node<T> first;
	protected Node<T> last;
	protected int size;

	public LinkedList(){
		size = 0;
	}

	public int size(){
		return size;
	}
	
	public boolean isEmpty(){
		return size == 0;
	}
	
	
	public T getFirst() throws IllegalStateException {
		if(isEmpty()) throw new IllegalStateException();

		return first.getValue();
	}
	
	public Node<T> getFirstNode() throws IllegalStateException {
		if(isEmpty()) throw new IllegalStateException();

		return first;
	}

	public T getLast() throws IllegalStateException {
		if(isEmpty()) throw new IllegalStateException();

		return last.getValue();
	}
	
	public Node<T> getLastNode() throws IllegalStateException {
		if(isEmpty()) throw new IllegalStateException();

		return last;
	}

	public void add(T value){
		addNode(new Node<T>(value));
		
	}
	
	public void addNode(Node<T> node){
		if (size == 0) {
			first = node;
		}
		else {
			last.setNext(node);
			node.setPrevious(last);
		}

		last = node;
		size++;
	}
	
	public void addFirst(T value){
		addNodeFirst(new Node<T>(value));
	}
	
	public void addNodeFirst(Node<T> node){
		if (!isEmpty()) {			
			first.setPrevious(node);
			node.setNext(first);
		}

		first = node;
		size++;
	}
	
	public void addLast(T value){
		add(value);
	}
	
	public void addNodeLast(Node<T> node){
		addNode(node);
	}
	
	public void addAll(T...args){
		for (int i = 0; i < args.length; i++) {
			add(args[i]);
		}
	}
	
	public void removeFirst(){
		if (!isEmpty()) {
			first = first.getNext();
			size--;
		}
	}
	
	public void removeLast(){
		if (!isEmpty()) {
			last = last.getPrevious();
			size--;
		}
	}
	
	public String toString (){
		String s = "[";
		
		if(size > 0){
			Node<T> node = first;
			
			while(node.hasNext()){
				s += node.getValue() + ", ";
				node = node.getNext();
			}
			
			s += node.getValue();
		}
		
		s += "]";
		
		return s;
	}
	
	public void print(){
		System.out.println(toString());
	}
	
	public static void main(String[] args) {
		LinkedList<Integer> l = new LinkedList<Integer>();
		
		l.addAll(1,2,3,4,5,6);
		l.addFirst(0);
		l.addLast(7);
		l.print();
		l.removeFirst();
		l.removeLast();
		l.print();
		l.removeFirst();
		l.print();
		
	}

}

File: js/src/linked-list/linked-list.js
// const LinkedListNode = require('./linked-list-node');
//
// module.exports = LinkedList;
//
// function LinkedList(){
//   this.first = null;
//   this.last = null;
//   this.size = 0;
// }
//
// LinkedList.prototype.getFirst = function () {
//   if (!this.size) {
//     return null;
//   }
//   return this.first.value;
// }
//
// LinkedList.prototype.getLast = function () {
//   if (!this.size) {
//     return null;
//   }
//   return this.last.value;
// }
//
// LinkedList.prototype.length = function () {
//   return this.size;
// };
//
// LinkedList.prototype.add = function (value) {
//   var newNode = new LinkedListNode({value: value});
//
//   if (this.first === null) {
//     this.first = newNode;
//     this.last = newNode;
//   }
//   else {
//     var oldLast = this.last;
//     oldLast.next = newNode;
//     this.last = newNode;
//     newNode.previous = oldLast;
//   }
//
//   this.size++;
// };
//
// LinkedList.prototype.addAll = function () {
//   for(var i = 0; i < arguments.length; i++){
//     this.add(arguments[i]);
//   }
// };
//
// LinkedList.prototype.addFirst = function (value) {
//   var newNode = new LinkedListNode({value: value});
//
//   if (this.first === null) {
//     this.first = newNode;
//     this.last = newNode;
//   }
//   else {
//     var oldFirst = this.first;
//     this.first = newNode;
//     newNode.next = oldFirst;
//   }
//
//   this.size++;
// };
//
// LinkedList.prototype.addLast = function (value) {
//   var newNode = new LinkedListNode({value: value});
//   var oldLast = this.last;
//
//   this.last = newNode;
//   oldLast.next = newNode;
//   newNode.previous = oldLast;
// };
//
// LinkedList.prototype.remove = function (index) {
//   var node = this.first;
//
//   if (index >= this.size) {
//     throw new Error('IndexOutOfRangeError');
//   }
//
//   for (var i = 0; i < index; i++) {
//     if (node.hasNext()) {
//       node = node.next;
//     }
//   }
//
//   if (node){
//     if(node !== this.first) {
//
//       if (node.hasNext()) {
//         node.previous.next = node.next;
//       }
//       else {
//         node.previous.next = null;
//         this.last = node.previous;
//       }
//
//     }
//     else {
//       if (this.size === 1) {
//         this.first = null;
//       }
//       else {
//         this.first = node.next
//       }
//
//     }
//
//     this.size--;
//   }
//
//   return node.value;
// };
//
// LinkedList.prototype.removeFirst = function () {
//   var first = this.first;
//
//   if (first.hasNext()) {
//       this.first = first.next;
//   }
//   else {
//     this.first = null;
//   }
//
//   this.size--;
//   return first.value;
// };
//
// LinkedList.prototype.removeLast = function () {
//   var last = this.last;
//   this.last = last.previous;
//   last.previous.next = null;
//
//   this.size--;
//   return last.value;
// };
//
// LinkedList.prototype.delete = function (node) {
//   if (node === this.first) {
//     this.first = null;
//     size--;
//   }
//   else {
//     node = node.next;
//     size--;
//   }
//
// };
//
// LinkedList.prototype.iterator = function(){
//   var self = this;
//   var curr = self.first;
//   var size = self.size;
//
//   return {
//     next: function(){
//       if (curr === null) {
//         return { done : true};
//       } else {
//         return {
//           value: function(){
//             var val = curr.value;
//             curr = curr.next;
//             return val;
//           }(),
//           done: false
//         }
//       }
//     }
//   }
// }
//
// LinkedList.prototype.contains = function (value) {
//   var curr = this.first
//
//   while(curr !== null){
//     if (curr.value === value) {
//       return true
//     }
//
//     return false;
//   }
// };
//
// LinkedList.prototype.toString = function () {
//   var node = this.first;
//   var s = '['
//
//   while (!(node === null)) {
//     if (node.hasNext()) {
//         s+= node.value + ', ';
//     }
//     else {
//       s += node.value;
//     }
//     node = node.next;
//   }
//
//   s+=']'
//   console.log(s);
//   return s;
// };
//
// if (require.main === module) {
//   var myLList = new LinkedList();
//   myLList.addAll(1,2,3,4,5,6,7,'are werr');
//   myLList.toString()
//   myLList.removeLast();
//   myLList.remove(5);
//   myLList.toString()
//   myLList.addFirst(0)
//   myLList.toString()
//   myLList.getFirst()
//   myLList.addLast('added last')
//   myLList.toString()
//   myLList.getLast()
//   console.log(myLList.size);
//   var i = myLList.iterator()
//   console.log(i.next().value);
//   console.log(i.next());
// }
